interfaces/builtin/i2c_control.go

```diff
@@ -0,0 +1,133 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/snapcore/snapd/interfaces"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+// The type for i2c control interface
+type I2cControlInterface struct{}
+
+// Getter for the name of the i2c-control interface
+func (iface *I2cControlInterface) Name() string {
+	return "i2c-control"
+}
+
+func (iface *I2cControlInterface) String() string {
+	return iface.Name()
+}
+
+// Pattern to match allowed i2c device nodes. It is gonna be used to check the
+// validity of the path attributes in case the udev is not used for
+// identification
+var i2cControlDeviceNodePattern = regexp.MustCompile("^/dev/i2c-[0-9]+$")
+
+// Check validity of the defined slot
+func (iface *I2cControlInterface) SanitizeSlot(slot *interfaces.Slot) error {
+
+	// Does it have right type?
+	if iface.Name() != slot.Interface {
+		panic(fmt.Sprintf("slot is not of interface %q", iface))
+	}
+
+	// Creation of the slot of this type
+	// is allowed only by a gadget snap
+	if slot.Snap.Type != "gadget" {
+		return fmt.Errorf("i2c-control slots only allowed on gadget snaps")
+	}
+
+	// Validate the path
+	path, ok := slot.Attrs["path"].(string)
+	if !ok || path == "" {
+		return fmt.Errorf("i2c-control slot must have a path attribute")
+	}
+
+	path = filepath.Clean(path)
+
+	if !i2cControlDeviceNodePattern.MatchString(path) {
+		return fmt.Errorf("i2c-control path attribute must be a valid device node")
+	}
+
+	return nil
+}
+
+// Checks and possibly modifies a plug
+func (iface *I2cControlInterface) SanitizePlug(plug *interfaces.Plug) error {
+	if iface.Name() != plug.Interface {
+		panic(fmt.Sprintf("plug is not of interface %q", iface))
+	}
+	// Currently nothing is checked on the plug side
+	return nil
+}
+
+// Returns snippet granted on install
+func (iface *I2cControlInterface) PermanentSlotSnippet(slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	return nil, nil
+}
+
+// Getter for the security snippet specific to the plug
+func (iface *I2cControlInterface) ConnectedPlugSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	path, pathOk := slot.Attrs["path"].(string)
+	if !pathOk {
+		return nil, nil
+	}
+	switch securitySystem {
+	case interfaces.SecurityAppArmor:
+		// No USB attributes for i2c so just go directly to
+		// considering the fixed node
+		cleanedPath := filepath.Clean(path)
+		return []byte(fmt.Sprintf("%s rw,\n", cleanedPath)), nil
+
+	case interfaces.SecurityUDev:
+		const udevRule string = `KERNEL="%s", TAG+="snap_%s_%s"`
```

Comment:
We agreed on allowing the access to a particular i2c device nodes instead of the bus as a whole for a single snap. Here it is tagged to explicitly move the /dev/i2c-whatever to the control group for this snap to have access to. Does it make sense?